WalletWasabi/Blockchain/TransactionBroadcasting/TransactionBroadcaster.cs

```diff
@@ -1,212 +1,225 @@
+		Logger.LogInfo($"Trying to broadcast transaction via backend API:{tx.GetHash()}.");
+		try
 		{
-			Logger.LogWarning($"Transaction {transaction.GetHash()} was already present in the broadcast store.");
+			var wasabiClient = new WasabiClient(wasabiHttpClientFactory.NewHttpClientWithCircuitPerRequest());
+			await wasabiClient.BroadcastAsync(tx).ConfigureAwait(false);
+			return BroadcastingResult.Ok();
 		}
-		var invPayload = new InvPayload(transaction.Transaction);
-
-		// Give 7 seconds to send the inv payload.
-		await node.SendMessageAsync(invPayload).WaitAsync(TimeSpan.FromSeconds(7)).ConfigureAwait(false); // ToDo: It's dangerous way to cancel. Implement proper cancellation to NBitcoin!
-
-		if (BitcoinStore.MempoolService.TryGetFromBroadcastStore(transaction.GetHash(), out TransactionBroadcastEntry? entry))
+		catch (HttpRequestException ex)
 		{
-			// Give 7 seconds for serving.
-			var timeout = 0;
-			while (!entry.IsBroadcasted())
+			if (RpcErrorTools.IsSpentError(ex.Message))
 			{
-				if (timeout > 7)
+				// If there is only one coin then that's the one that is already spent (what about full-RBF?).
+				if (tx.Transaction.Inputs.Count == 1)
 				{
-					throw new TimeoutException("Did not serve the transaction.");
+					OutPoint input = tx.Transaction.Inputs[0].PrevOut;
+					return BroadcastingResult.Fail(new BroadcastError.SpentInputError(input));
 				}
-				await Task.Delay(1_000).ConfigureAwait(false);
-				timeout++;
-			}
-			node.DisconnectAsync("Thank you!");
-			Logger.LogInfo($"Disconnected node: {node.RemoteSocketAddress}. Successfully broadcasted transaction: {transaction.GetHash()}.");
 
-			// Give 21 seconds for propagation.
-			timeout = 0;
-			while (entry.GetPropagationConfirmations() < 2)
-			{
-				if (timeout > 21)
-				{
-					throw new TimeoutException("Did not serve the transaction.");
-				}
-				await Task.Delay(1_000).ConfigureAwait(false);
-				timeout++;
+				return BroadcastingResult.Fail(new BroadcastError.SpentError());
 			}
-			Logger.LogInfo($"Transaction is successfully propagated: {transaction.GetHash()}.");
+			return BroadcastingResult.Fail(new BroadcastError.Unknown(ex.Message));
 		}
-		else
+	}
+}
+
+public class NetworkBroadcaster(MempoolService mempoolService, NodesGroup nodes) : IBroadcaster
+{
+	public async Task<BroadcastingResult> BroadcastAsync(SmartTransaction tx)
+	{
+		if (nodes.ConnectedNodes.Count < 2)
 		{
-			Logger.LogWarning($"Expected transaction {transaction.GetHash()} was not found in the broadcast store.");
+			return BroadcastingResult.Fail(new BroadcastError.NotEnoughP2pNodes());
 		}
+		var connectedNodeCount = nodes.ConnectedNodes.Count;
+		var broadcastToNodes = nodes.ConnectedNodes
+			.Where(n => n.IsConnected)
+			.OrderBy(_ => Guid.NewGuid())
+			.Take(1 + connectedNodeCount / 3)
+			.Select(n => BroadcastCoreAsync(tx, n))
+			.ToArray();
+
+		var results = await Task.WhenAll(broadcastToNodes).ConfigureAwait(false);
```

Comment:
Can't we use more nodes, use `.WhenAny`, and upon first success return but keep the rest of the tasks going on in the background to improve propagation?

That way we could increase the timeout safely.

Instead of just one success it might also be safer to wait for 2 or 3 successes.

Because the way it works with your PR doesn't sound so wise as it waits for all nodes to either success either fail (or timeout to kick in) but then all results except first success are ignored